File: solune/backend/src/models/chores.py
"""Pydantic models for Chores — recurring maintenance tasks."""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, Field

# ── Enums ──


class ScheduleType(StrEnum):
    """Chore schedule types."""

    TIME = "time"
    COUNT = "count"


class ChoreStatus(StrEnum):
    """Chore status values."""

    ACTIVE = "active"
    PAUSED = "paused"


# ── Core Models ──


class Chore(BaseModel):
    """Full chore record returned from the API."""

    id: str
    project_id: str
    name: str
    template_path: str
    template_content: str
    schedule_type: ScheduleType | None = None
    schedule_value: int | None = None
    status: ChoreStatus = ChoreStatus.ACTIVE
    last_triggered_at: str | None = None
    last_triggered_count: int = 0
    current_issue_number: int | None = None
    current_issue_node_id: str | None = None
    pr_number: int | None = None
    pr_url: str | None = None
    tracking_issue_number: int | None = None
    execution_count: int = 0
    ai_enhance_enabled: bool = True
    agent_pipeline_id: str = ""
    is_preset: bool = False
    preset_id: str = ""
    created_at: str
    updated_at: str


class ChoreCreate(BaseModel):
    """Request body for creating a new chore."""

    name: str = Field(..., min_length=1, max_length=200)
    template_content: str = Field(..., min_length=1)


class ChoreUpdate(BaseModel):
    """Request body for updating a chore (partial)."""

    schedule_type: ScheduleType | None = None
    schedule_value: int | None = Field(default=None, gt=0)
    status: ChoreStatus | None = None
    ai_enhance_enabled: bool | None = None
    agent_pipeline_id: str | None = None


# ── Trigger Models ──


class ChoreTriggerResult(BaseModel):
    """Result of triggering (or attempting to trigger) a single chore."""

    chore_id: str
    chore_name: str
    triggered: bool
    issue_number: int | None = None
    issue_url: str | None = None
    skip_reason: str | None = None


class EvaluateChoreTriggersResponse(BaseModel):
    """Response for the evaluate-triggers endpoint."""

    evaluated: int
    triggered: int
    skipped: int
    results: list[ChoreTriggerResult]


class EvaluateChoreTriggersRequest(BaseModel):
    """Optional request body for the evaluate-triggers endpoint."""

    project_id: str | None = None
    parent_issue_count: int | None = None


class TriggerChoreRequest(BaseModel):
    """Optional request body for the manual trigger endpoint."""

    parent_issue_count: int | None = None


# ── Template Models ──


class ChoreTemplate(BaseModel):
    """A chore template discovered from .github/ISSUE_TEMPLATE/."""

    name: str
    about: str
    path: str
    content: str


# ── Chat Models ──


class ChoreChatMessage(BaseModel):
    """Request body for the chore chat endpoint."""

    content: str
    conversation_id: str | None = None
    ai_enhance: bool = True


class ChoreChatResponse(BaseModel):
    """Response from the chore chat endpoint."""

    message: str
    conversation_id: str
    template_ready: bool = False
    template_content: str | None = None
    template_name: str | None = None


# ── Inline Editing Models ──


class ChoreInlineUpdate(BaseModel):
    """Request body for inline editing of a Chore definition.

    A PR is only created when ``name`` or ``template_content`` is set.
    All other fields are config-only and saved directly without creating a PR.
    """

    name: str | None = Field(default=None, min_length=1, max_length=200)
    template_content: str | None = Field(default=None, min_length=1)
    schedule_type: ScheduleType | None = None
    schedule_value: int | None = Field(default=None, gt=0)
    ai_enhance_enabled: bool | None = None
    agent_pipeline_id: str | None = None
    expected_sha: str | None = None


class ChoreInlineUpdateResponse(BaseModel):
    """Response from inline Chore update with PR creation result."""

    chore: Chore
    pr_number: int | None = None
    pr_url: str | None = None
    pr_merged: bool = False
    merge_error: str | None = None


# ── Creation with Auto-Merge Models ──


class ChoreCreateWithConfirmation(BaseModel):
    """Request body for creating a new Chore with the full auto-merge flow."""

    name: str = Field(..., min_length=1, max_length=200)
    template_content: str = Field(..., min_length=1)
    ai_enhance_enabled: bool = True
    agent_pipeline_id: str = ""
    auto_merge: bool = True


class ChoreCreateResponse(BaseModel):
    """Response from Chore creation with full flow result."""

    chore: Chore
    issue_number: int | None = None
    pr_number: int | None = None
    pr_url: str | None = None
    pr_merged: bool = False
    merge_error: str | None = None

File: solune/frontend/src/types/chores.ts
/**
 * Chore domain types.
 */

// ============ Chores Types (016-replace-housekeeping-chores) ============

export type ScheduleType = 'time' | 'count';
export type ChoreStatus = 'active' | 'paused';

export interface Chore {
  id: string;
  project_id: string;
  name: string;
  template_path: string;
  template_content: string;
  schedule_type: ScheduleType | null;
  schedule_value: number | null;
  status: ChoreStatus;
  last_triggered_at: string | null;
  last_triggered_count: number;
  current_issue_number: number | null;
  current_issue_node_id: string | null;
  pr_number: number | null;
  pr_url: string | null;
  tracking_issue_number: number | null;
  execution_count: number;
  ai_enhance_enabled: boolean;
  agent_pipeline_id: string;
  is_preset: boolean;
  preset_id: string;
  created_at: string;
  updated_at: string;
}

export interface ChoreCreate {
  name: string;
  template_content: string;
}

export interface ChoreTemplate {
  name: string;
  about: string;
  path: string;
  content: string;
}

export interface ChoreUpdate {
  schedule_type?: ScheduleType | null;
  schedule_value?: number | null;
  status?: ChoreStatus;
  ai_enhance_enabled?: boolean;
  agent_pipeline_id?: string;
}

export interface ChoreTriggerResult {
  chore_id: string;
  chore_name: string;
  triggered: boolean;
  issue_number: number | null;
  issue_url: string | null;
  skip_reason: string | null;
}

export interface EvaluateChoreTriggersResponse {
  evaluated: number;
  triggered: number;
  skipped: number;
  results: ChoreTriggerResult[];
}

export interface ChoreChatMessage {
  content: string;
  conversation_id?: string | null;
  ai_enhance?: boolean;
}

export interface ChoreChatResponse {
  message: string;
  conversation_id: string;
  template_ready: boolean;
  template_content: string | null;
  template_name: string | null;
}

// ── Inline Editing Types ──

export interface ChoreInlineUpdate {
  name?: string;
  template_content?: string;
  schedule_type?: ScheduleType | null;
  schedule_value?: number | null;
  ai_enhance_enabled?: boolean;
  agent_pipeline_id?: string;
  expected_sha?: string;
}

export interface ChoreInlineUpdateResponse {
  chore: Chore;
  pr_number: number | null;
  pr_url: string | null;
  pr_merged: boolean;
  merge_error: string | null;
}

export interface ChoreCreateWithConfirmation {
  name: string;
  template_content: string;
  ai_enhance_enabled: boolean;
  agent_pipeline_id: string;
  auto_merge: boolean;
}

export interface ChoreCreateResponse {
  chore: Chore;
  issue_number: number | null;
  pr_number: number | null;
  pr_url: string | null;
  pr_merged: boolean;
  merge_error: string | null;
}

// ── Featured Rituals Types ──

export interface FeaturedRitualCard {
  choreId: string;
  choreName: string;
  stat: string;
  statValue: number;
}

export interface FeaturedRituals {
  nextRun: FeaturedRitualCard | null;
  mostRecentlyRun: FeaturedRitualCard | null;
  mostRun: FeaturedRitualCard | null;
}

export interface ChoreEditState {
  original: Chore;
  current: Partial<ChoreInlineUpdate>;
  isDirty: boolean;
  fileSha: string | null;
}

export interface ChoreCounterData {
  choreId: string;
  remaining: number;
  totalThreshold: number;
  issuesSinceLastRun: number;
}
